File: 236-LowestCommonAncestorOfABinaryTree/236-LowestCommonAncestorOfABinaryTree.cpp
// Last updated: 08/05/2026, 09:30:17
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:

    TreeNode* dfs(TreeNode* root, TreeNode* &p, TreeNode* &q){
        if(!root){
            return nullptr;
        }
        if(root==p){
            return p;
        }
        else if(root==q){
            return q;
        }
        TreeNode* left= dfs(root->left,p,q );
        TreeNode* right= dfs(root->right,p,q);

        if(left && right){
            return root;
        }
        else if(!left){
            return right;
        }
        else{
            return left;
        }

        return nullptr;
    }

    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        return dfs(root,p,q);
    }
};

File: 236-LowestCommonAncestorOfABinaryTree/236-LowestCommonAncestorOfABinaryTree.py
# Last updated: 08/05/2026, 09:30:23
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':

        def solve(node: TreeNode):
            
            if node== None or node==p or node==q:
                return node

            left = solve(node.left)
            right = solve(node.right)

            if left== None:
                return right
            elif right== None:
                return left
            else:
                return node

        return solve(root)


